kratos/python/add_constitutive_law_to_python.cpp

```diff
@@ -91,10 +91,31 @@ void DeprecatedInitializeSolutionStep( ConstitutiveLaw& rThisConstitutiveLaw,
     // function wrapper to issue a deprecation-warning when called from python
     KRATOS_WARNING_FIRST_N("ConstitutiveLaw-Python Interface", 10) << "\"InitializeSolutionStep\" "
         << "is deprecated, please don't use it" << std::endl;
+
+    #if defined(__clang__)
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wdeprecated-declarations"
+    #elif defined(__GNUC__) || defined(__GNUG__)
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+    #elif defined(_MSC_VER)
+    #pragma warning(push)
+    #pragma warning(disable: 4996)
+    #endif
+
     rThisConstitutiveLaw.InitializeSolutionStep(rMaterialProperties,
                                                 rElementGeometry,
                                                 rShapeFunctionsValues,
                                                 rCurrentProcessInfo);
+
+    #if defined(__clang__)
+    #pragma clang diagnostic pop
+    #elif defined(__GNUC__) || defined(__GNUG__)
+    #pragma GCC diagnostic pop
+    #elif defined(_MSC_VER)
+    #pragma warning(pop)
+    #endif
```

Comment:
why is this needed?

Reply:
Right now, when we compile the core, everyone gets a warning saying that the method InitializeSolutionStep is deprecated. I think it's not nice to show this warning to everyone, specially to those who don't use this method or don't care about it. However, this method is exported to Python, so it means that the core contains a call to this deprecated method. 
For calls from Python, a KRATOS_WARNING appears in runtime (OK). For calls from c++ (applications) a compiler warning appears (OK). But for this single call from the core, I am trying to hide the warning, and this is what this implementation does, it deactivates the 'deprecated' warning just for one line of the core.

Reply:
hm, I would rather go for removing this function for good (since it is deprecated) 
@RiccardoRossi you added this, what do you think?

I am a bit afraid that if we start to do this also others will start doing it, which is against what it is intended to do

After all, it is just a warning, as you write, nothing to concern you if you are not using it
Why is this a problem for you?

Reply:
I see, I also don't like warnings :) Yet deprecating things seems to be a better way than breaking things. I think also python/numpy is doing it this way
Not sure if it is possible though to always have a clean core compilation, since a lot of things are there

Different approach:
We could add `-Wno-deprecated-declarations` to the standard-flags, then users won't see those warnings at all
Doesn't change the situation for developers though...

If the code you are adding is necessary, then we could create a macro for it, which can then also be used in similar situations. Ofc a comment has to be added that this macro should rarely be used
~~~c
#define KRATOS_DISABLE_DEPRECATION_WARNING
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
#elif defined(__GNUC__) || defined(__GNUG__)
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
#elif defined(_MSC_VER)
#pragma warning(push)
#pragma warning(disable: 4996)
#endif
~~~

Reply:
I would create two macros having all these definition but showing the intention:

```c++
KRATOS_CALL_DEPRECATED_FUNTION_BEGIN
    rThisConstitutiveLaw.InitializeSolutionStep(rMaterialProperties,
                                                rElementGeometry,
                                                rElementGeometry,
                                                rShapeFunctionsValues,
                                                rShapeFunctionsValues,
                                                rCurrentProcessInfo);
                                                rCurrentProcessInfo);
KRATOS_CALL_DEPRECATED_FUNTION_END
```
And it would be more searchable for future removing

Reply:
I thought about this as well, but I didn't find a way to code a macro with pre-processor sentences inside. Maybe I should write all of this in a function? @roigcarlo told me that he could not think of a nice solution in this spot.

Reply:
My idea was something like this but I'm not sure if it works:

```c++
#define KRATOS_DISABLE_DEPRECATION_WARNING
#if defined(__clang__)
#define KRATOS_CALL_DEPRECATED_FUNTION_BEGIN \
_Pragma(clang diagnostic push) \
_Pragma(clang diagnostic ignored "-Wdeprecated-declarations")
#elif defined(__GNUC__) || defined(__GNUG__)
#define KRATOS_CALL_DEPRECATED_FUNTION_BEGIN \
_Pragma(GCC diagnostic push)\
_Pragma(GCC diagnostic ignored "-Wdeprecated-declarations")
#elif defined(_MSC_VER)
#define KRATOS_CALL_DEPRECATED_FUNTION_BEGIN \
__pragma(warning(push))\
__pragma(warning(disable: 4996))
#endif
```

Reply:
Thanks @pooyan-dadvand , nice approach! Finally I managed to compile. Note that I needed to create KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION because there are nested literals (_Pragma needs to get an argument between quote marks, but there were two levels of that: "GCC diagnostic ignored "-Wdeprecated-declarations" ").

```
#if defined(__clang__)
#define KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(x) _Pragma(#x)
#define KRATOS_START_IGNORING_DEPRECATED_FUNCTION_WARNING \
KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(clang diagnostic push) \
KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(clang diagnostic ignored "-Wdeprecated-declarations")
#elif defined(__GNUC__) || defined(__GNUG__)
#define KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(x) _Pragma(#x)
#define KRATOS_START_IGNORING_DEPRECATED_FUNCTION_WARNING \
KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(GCC diagnostic push) \
KRATOS_PRAGMA_INSIDE_MACRO_DEFINITION(GCC diagnostic ignored "-Wdeprecated-declarations")
#elif defined(_MSC_VER)
#define KRATOS_START_IGNORING_DEPRECATED_FUNCTION_WARNING \
__pragma(warning(push))\
__pragma(warning(disable: 4996))
#endif

#if defined(__clang__)
#define KRATOS_STOP_IGNORING_DEPRECATED_FUNCTION_WARNING \
_Pragma("clang diagnostic pop")
#elif defined(__GNUC__) || defined(__GNUG__)
#define KRATOS_STOP_IGNORING_DEPRECATED_FUNCTION_WARNING \
_Pragma("GCC diagnostic pop")
#elif defined(_MSC_VER)
#define KRATOS_STOP_IGNORING_DEPRECATED_FUNCTION_WARNING \
__pragma(warning(pop))
#endif
```

Reply:
More complicated than what I would imagine! Good macro name and good job!